File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # (leftMost, height)
        stack = []
        res = -1
        for i, height in enumerate(heights):
            start = i
            while stack and stack[-1][1] > height:
                index, h = stack.pop()
                res = max(res, h * (i - index))
                start = index
            stack.append((start, height))
        
        # now find the max
        for index, height in stack:
            res = max(res, height * (len(heights) - index))
        return res

File: Data Structures & Algorithms/palindromic-substrings/submission-0.cpp
class Solution {
public:
    int countSubstrings(const string& s) {
        int n = s.size();
        vector<vector<bool>> check(n, vector<bool>(n, false));
        int res = 0;
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s[i] == s[j] && ((j - i) <= 2 || check[i + 1][j - 1])) {
                    res++;
                    check[i][j] = true;
                }
            }
        }
        return res;
    }
};
